### PR TITLE
WT-9115 Accept EINVAL error code from commit in the cpp suite

### PR DESCRIPTION
**Pull request by etienneptl**
Since the timestamp manager moves periodically the oldest and stable timestamps, sometimes we end up having transactions with a commit timestamp behind those timestamps. WiredTiger considers this as invalid and returns EINVAL. 

- We now check the transaction still has a valid commit timestamp before committing it and call rollback if that's not the case.
- Now the test can get further, it generates too much data and fills up the disk space of the host which executes it. We disabled the tracking component of the `operation_test` test when the stressing configuration is used. This will be fixed in WT-8640.
- Finally, some code cleanup was done.

### CODE REVIEW

File: test/cppsuite/test_harness/workload/thread_context.cpp [hunk 1]
```diff
@@ -109,6 +109,19 @@ transaction_context::commit(const std::string &config)
 {
     WT_DECL_RET;
     testutil_assert(_in_txn && !_needs_rollback);
+
+    /*
+     * We may have set the commit ts to a value that is now older than the current stable timestamp,
+     * we need to rollback as it is an invalid transaction. TODO - Should we have this check here or
```
Comment (etienneptl) [inline test/cppsuite/test_harness/workload/thread_context.cpp:115]: Happy to get some ideas here regarding the `TODO`.

File: test/cppsuite/test_harness/workload/thread_context.cpp [hunk 2]
```diff
@@ -109,6 +109,19 @@ transaction_context::commit(const std::string &config)
...
+     * in the different operations defined in the database_operation class?
+     */
+    if (_timestamp_manager->get_stable_ts() >
+      ((WT_SESSION_IMPL *)_session)->txn->commit_timestamp) {
```
Comment (etienneptl) [inline test/cppsuite/test_harness/workload/thread_context.cpp:119]: Should we use `__session_query_timestamp` instead of accessing the `commit_timestamp` field directly?

**Comment by etienneptl** [comment]
I am setting this PR as a draft as we are still deciding on the solution for this issue.

**Comment by etienneptl** [comment]
Ready for review again, as discussed in WT-9115, this solution is ok for now.

File: test/cppsuite/test_harness/workload/thread_context.cpp (hunk 2 above)
Comment (etienneptl) [reply test/cppsuite/test_harness/workload/thread_context.cpp:119]: I don't think we have to do anything here, the code is basically the same as in [__txn_query_timestamp](https://github.com/wiredtiger/wiredtiger/blob/develop/src/txn/txn_timestamp.c#L224):
```
txn = session->txn;
...
*tsp = txn->commit_timestamp;
```

File: test/cppsuite/test_harness/workload/thread_context.cpp (hunk 1 above)
Comment (ajmorton) [reply test/cppsuite/test_harness/workload/thread_context.cpp:117]: I'm in favour of a single common check here. I can't see this behaviour needing to differ across operations.

File: test/cppsuite/test_harness/workload/thread_context.cpp [hunk 3]
```diff
@@ -111,6 +111,19 @@ transaction_context::commit(const std::string &config)
...
+        logger::log_msg(LOG_WARN, "Failed to commit transaction in commit due to invalid ts");
```
Comment (ajmorton) [inline test/cppsuite/test_harness/workload/thread_context.cpp:122]: Recommended reword: "Rolling back transaction as its commit_ts is less than stable_ts"

File: test/cppsuite/test_harness/workload/thread_context.cpp [hunk 4]
```diff
@@ -111,6 +111,19 @@ transaction_context::commit(const std::string &config)
 {
     WT_DECL_RET;
     testutil_assert(_in_txn && !_needs_rollback);
+
+    /*
```
Comment (ajmorton) [inline test/cppsuite/test_harness/workload/thread_context.cpp:115]: Do we need a FIXME to remove this section on completion of WT-9198?

**Review by ajmorton — APPROVED** [review]
LGTM. Some questions, no blockers.

File: test/cppsuite/test_harness/workload/thread_context.cpp (hunk 1 above)
Comment (jiechenbo) [reply test/cppsuite/test_harness/workload/thread_context.cpp:117]: I'm with @ajmorton here

File: test/cppsuite/test_harness/workload/thread_context.cpp (hunk 2 above)
Comment (luke-pearson) [reply test/cppsuite/test_harness/workload/thread_context.cpp:121]: This check will begin failing as soon as the first_commit_timestamp is used instead of the commit_timestamp. Also casting the session to a WT_SESSION_IMPL outside the api is a big no-no. One option to track this would be to save the first commit timestamp used on the thread context and then check against the stable. However the stable can still move concurrently and then you'd receive EINVAL in commit. So I think we should just check for EINVAL and rollback as required until we can implement a proper solution.

File: test/cppsuite/test_harness/workload/thread_context.cpp [hunk 5]
```diff
@@ -240,30 +253,23 @@ thread_context::update(scoped_cursor &cursor, uint64_t collection_id, const std:
...
+thread_context::insert(scoped_cursor &cursor, uint64_t collection_id, uint64_t key_id)
 {
-    return insert(cursor, collection_id, key_to_string(key_id), ts);
+    return insert(cursor, collection_id, key_to_string(key_id));
 }
 
 bool
-thread_context::insert(
-  scoped_cursor &cursor, uint64_t collection_id, const std::string &key, wt_timestamp_t ts)
+thread_context::insert(scoped_cursor &cursor, uint64_t collection_id, const std::string &key)
```
[the remark this answers is not in the document]
Comment (etienneptl) [reply test/cppsuite/test_harness/workload/thread_context.cpp:262]: I thought so too! But I could not find why we had those, I am happy to revert the changes as it is not really related to this ticket and create a new one where we could investigate more.
What I found so far is that we introduced the `ts` argument as part of https://github.com/wiredtiger/wiredtiger/pull/7006 in this [commit](https://github.com/wiredtiger/wiredtiger/pull/7006/commits/84e5ceb20ceed14f8050d7eb776a3fd379c82d22).

File: test/cppsuite/test_harness/workload/thread_context.cpp (hunk 2 above)
Comment (etienneptl) [reply test/cppsuite/test_harness/workload/thread_context.cpp:121]: @jiechenbo, no need to check for the oldest ts as by definition it is older than the stable ts, I will add a comment for this!
@luke-pearson, I will try this approach.

File: test/cppsuite/test_harness/workload/thread_context.cpp (hunk 1 above)
Comment (etienneptl) [reply test/cppsuite/test_harness/workload/thread_context.cpp]: Thanks, updated in efd54370a5f8fae9703829ea1ffb39382a5ba6c8

File: test/cppsuite/test_harness/workload/thread_context.cpp (hunk 2 above)
Comment (etienneptl) [reply test/cppsuite/test_harness/workload/thread_context.cpp]: Remove this entirely as we are accepting EINVAL now until WT-9198 is done.

File: test/cppsuite/test_harness/workload/thread_context.cpp (hunk 3 above)
Comment (etienneptl) [reply test/cppsuite/test_harness/workload/thread_context.cpp]: See efd54370a5f8fae9703829ea1ffb39382a5ba6c8

File: test/cppsuite/test_harness/workload/thread_context.cpp (hunk 4 above)
Comment (etienneptl) [reply test/cppsuite/test_harness/workload/thread_context.cpp:116]: See efd54370a5f8fae9703829ea1ffb39382a5ba6c8

File: test/cppsuite/test_harness/workload/thread_context.h [hunk 6]
```diff
@@ -135,10 +135,8 @@ class thread_context {
      * Return true if the operation was successful, a return value of false implies the transaction
      * needs to be rolled back.
      */
-    bool insert(
-      scoped_cursor &cursor, uint64_t collection_id, uint64_t key_id, wt_timestamp_t ts = 0);
-    bool insert(
-      scoped_cursor &cursor, uint64_t collection_id, const std::string &key, wt_timestamp_t ts = 0);
+    bool insert(scoped_cursor &cursor, uint64_t collection_id, uint64_t key_id);
+    bool insert(scoped_cursor &cursor, uint64_t collection_id, const std::string &key);
```
Comment (AlexeyAnisimovAU) [inline test/cppsuite/test_harness/workload/thread_context.h:139]: Why did you remove timestamps from inserts?
Also if timestamps are gone, the comment for the methods needs to be updated.

Comment (etienneptl) [reply test/cppsuite/test_harness/workload/thread_context.h:138]: @AlexeyAnisimovAU, it is not used anywhere, some code cleanup.
Thanks for spotting the incorrect comment, updated in 4ff102d485eb4b260573116f0ba5602b7cc15da6.

**Review by jiechenbo — APPROVED** [review]
LGTM!

File: test/cppsuite/test_harness/workload/thread_context.cpp (hunk 2 above)
Comment (etienneptl) [reply test/cppsuite/test_harness/workload/thread_context.cpp]: Furthermore, we don't need to call rollback at the cpp layer as WT will take care of it internally.

**Review by AlexeyAnisimovAU — APPROVED** [review]
LGTM!
